src/sql/workbench/services/query/common/gridDataProvider.ts

```diff
@@ -45,94 +49,144 @@ export interface IGridDataProvider {
-			let columnHeaders = provider.getColumnHeaders(range);
-			if (columnHeaders !== undefined) {
-				let idx = 0;
-				for (let header of columnHeaders) {
-					headers.set(startCol + idx, header);
-					idx++;
-				}
-			}
-			// Iterate over the rows to paste into the copy string
-			for (let rowIndex: number = 0; rowIndex < result.length; rowIndex++) {
-				let row = result[rowIndex];
-				let cellObjects = row.slice(range.fromCell, (range.toCell + 1));
-				// Remove newlines if requested
-				let cells = provider.shouldRemoveNewLines()
-					? cellObjects.map(x => removeNewLines(x.displayValue))
-					: cellObjects.map(x => x.displayValue);
-
-				let idx = 0;
-				for (let cell of cells) {
-					let map = rows.get(rowIndex + startRow);
-					if (!map) {
-						map = new Map();
-						rows.set(rowIndex + startRow, map);
-					}
+interface Range {
+	start: number;
+	end: number;
+}
 
-					map.set(startCol + idx, cell);
-					idx++;
-				}
+/**
+ * Merge the ranges and get the sorted ranges.
+ */
+function mergeRanges(ranges: Range[]): Range[] {
+	const mergedRanges: Range[] = [];
+	const orderedRanges = ranges.sort((s1, s2) => { return s1.start - s2.start; });
+	orderedRanges.forEach(range => {
+		let merged = false;
+		for (let i = 0; i < mergedRanges.length; i++) {
+			const mergedRange = mergedRanges[i];
+			if (range.start <= mergedRange.end) {
+				mergedRange.end = Math.max(range.end, mergedRange.end);
+				merged = true;
+				break;
 			}
-		};
+		}
+		if (!merged) {
+			mergedRanges.push(range);
+		}
 	});
+	return mergedRanges;
+}
 
-	// Set the tasks gathered above to execute
-	let actionedTasks: Promise<void>[] = tasks.map(t => { return t(); });
+export async function copySelectionToClipboard(clipboardService: IClipboardService, notificationService: INotificationService, provider: IGridDataProvider, selections: Slick.Range[], includeHeaders?: boolean, tableView?: IDisposableDataProvider<Slick.SlickData>): Promise<void> {
+	const batchSize = 100;
+	const eol = provider.getEolString();
+	const valueSeparator = '\t';
+	const shouldRemoveNewLines = provider.shouldRemoveNewLines();
 
-	// Make sure all these tasks have executed
-	await Promise.all(actionedTasks);
+	// Merge the selections to get the columns and rows.
+	const columnRanges: Range[] = mergeRanges(selections.map(selection => { return { start: selection.fromCell, end: selection.toCell }; }));
+	const rowRanges: Range[] = mergeRanges(selections.map(selection => { return { start: selection.fromRow, end: selection.toRow }; }));
 
-	headers = sortMapEntriesByColumnOrder(headers);
-	rows = sortMapEntriesByColumnOrder(rows);
+	const totalRows = rowRanges.map(range => range.end - range.start + 1).reduce((p, c) => p + c);
 
-	let copyString = '';
+	let processedRows = 0;
+	const getMessageText = (): string => {
+		return nls.localize('gridDataProvider.loadingRowsInProgress', "Loading the rows to be copied({0}/{1})...", processedRows, totalRows);
```

Comment:
nit - it would look better to have a space before the parens

```suggestion
		return nls.localize('gridDataProvider.loadingRowsInProgress', "Loading the rows to be copied ({0}/{1})...", processedRows, totalRows);
```